compiler/modules/front/stmt.bal

```diff
@@ -911,7 +911,7 @@ function lookupVarRefForAssign(StmtContext cx, Environment env, string varName,
 
 function codeGenAssign(StmtContext cx, Environment env, bir:BasicBlock block, bir:Register result, s:Expr expr, t:SemType semType, Position pos) returns CodeGenError|bir:BasicBlock {
     var { result: operand, block: nextBlock } = check cx.codeGenExpr(block, env, semType, expr);
-    if result !is bir:TmpRegister|bir:VarRegister|bir:FinalRegister {
+    if result !is bir:AssignTmpRegister|bir:VarRegister|bir:FinalRegister {
```

Comment:
We should be passing in the right kind of register here not casting. This means more changes are still needed. I guess Binding needs changing to use the right kind of register.

Reply:
We currently use bindings with register to which you can't assign values at 2 places

1. https://github.com/ballerina-platform/nballerina/blob/1db49a0f695ac02032f4e8556e80f212dd6ae89e/compiler/modules/front/stmt.bal#L226
2. https://github.com/ballerina-platform/nballerina/blob/1db49a0f695ac02032f4e8556e80f212dd6ae89e/compiler/modules/front/stmt.bal#L801

If we are to keep allowing this we can also check for register type in [`lookupVarRefForAssign`](https://github.com/ballerina-platform/nballerina/blob/1db49a0f695ac02032f4e8556e80f212dd6ae89e/compiler/modules/front/stmt.bal#L890) and return the correct type (or panic if it is not the correct type)

Reply:
The binding will need to allow for
- params
- final variables
- narrowed variables
- normal variables

Only the last one can be assigned to.

The first two correspond to errors in the source: so these cases should be rejected during the lookup as a semantic error. In the third case, the assignment will cause the variable not be narrowed.  There will be a different type for the original binding, which won't allowed for narrowed registers.

There's no need to explicitly check and panic: just do the cast and put a comment explaining why the cast must succeed.

Reply:
fixed